clippy_lints/src/implicit_return.rs

```diff
@@ -0,0 +1,131 @@
+                    Self::expr_match(cx, expr);
+                }
+                // only needed in the case of `break` with `;` at the end
+                else if let Some(stmt) = block.stmts.last() {
+                    if let rustc::hir::StmtKind::Semi(expr, ..) = &stmt.node {
+                        Self::expr_match(cx, expr);
+                    }
+                }
+            },
+            // use `return` instead of `break`
+            ExprKind::Break(.., break_expr) => {
+                if let Some(break_expr) = break_expr {
+                    span_lint_and_then(cx, IMPLICIT_RETURN, expr.span, "missing return statement", |db| {
+                        if let Some(snippet) = snippet_opt(cx, break_expr.span) {
+                            db.span_suggestion_with_applicability(
+                                expr.span,
+                                "change `break` to `return` as shown",
+                                format!("return {}", snippet),
+                                Applicability::MachineApplicable,
+                            );
+                        }
+                    });
+                }
+            },
+            ExprKind::If(.., if_expr, else_expr) => {
+                Self::expr_match(cx, if_expr);
+
+                if let Some(else_expr) = else_expr {
+                    Self::expr_match(cx, else_expr);
+                }
+            },
+            ExprKind::Match(_, arms, ..) => {
+                for arm in arms {
+                    Self::expr_match(cx, &arm.body);
+                }
+            },
+            // loops could be using `break` instead of `return`
+            ExprKind::Loop(block, ..) => {
+                if let Some(expr) = &block.expr {
+                    Self::expr_match(cx, expr);
+                }
+            },
+            // skip if it already has a return statement
+            ExprKind::Ret(..) => (),
+            // everything else is missing `return`
+            _ => span_lint_and_then(cx, IMPLICIT_RETURN, expr.span, "missing return statement", |db| {
+                if let Some(snippet) = snippet_opt(cx, expr.span) {
+                    db.span_suggestion_with_applicability(
+                        expr.span,
+                        "add `return` as shown",
+                        format!("return {}", snippet),
+                        Applicability::MachineApplicable,
+                    );
+                }
+            }),
+        }
+    }
+}
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(IMPLICIT_RETURN)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'a, 'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl,
+        body: &'tcx Body,
+        _: Span,
+        _: NodeId,
+    ) {
+        let def_id = cx.tcx.hir.body_owner_def_id(body.id());
+        let mir = cx.tcx.optimized_mir(def_id);
```

Comment:
There's no need to invoke the MIR here. You should be able to get the return type from the `FnDecl`

Reply:
That doesn't work sadly. Otherwise I could have done the whole thing in `EarlyLintPass` and `syntax::ast`.
The issue is that I can't inferred closure return types and therefore can't determine if the last statement needs a `return` or is just `()`.
Unless there is something I'm missing?

Reply:
I added a comment to explain it in https://github.com/rust-lang/rust-clippy/pull/3494/commits/b0f3ed2b808d0696d0a97173d65d368b01c0c9a7.

Reply:
An early lint pass won't work, because types aren't known yet.

I guess we can leave it as it is, as all other methods also go through a query